reconcile/vault_replication.py

```diff
@@ -0,0 +1,392 @@
+import logging
+import re
+from collections.abc import Iterable
+from typing import (
+    Optional,
+    Union,
+    cast,
+)
+
+from reconcile.gql_definitions.jenkins_configs import jenkins_configs
+from reconcile.gql_definitions.jenkins_configs.jenkins_configs import (
+    JenkinsConfigsQueryData,
+    JenkinsConfigV1_JenkinsConfigV1,
+)
+from reconcile.gql_definitions.vault_instances import vault_instances
+from reconcile.gql_definitions.vault_instances.vault_instances import (
+    VaultInstanceV1_VaultReplicationConfigV1_VaultInstanceAuthV1,
+    VaultInstanceV1_VaultReplicationConfigV1_VaultInstanceAuthV1_VaultInstanceAuthApproleV1,
+    VaultReplicationConfigV1,
+    VaultReplicationConfigV1_VaultInstanceAuthV1,
+    VaultReplicationConfigV1_VaultInstanceAuthV1_VaultInstanceAuthApproleV1,
+    VaultReplicationJenkinsV1,
+)
+from reconcile.gql_definitions.vault_policies import vault_policies
+from reconcile.gql_definitions.vault_policies.vault_policies import (
+    VaultPoliciesQueryData,
+)
+from reconcile.utils import gql
+from reconcile.utils.vault import (
+    SecretAccessForbidden,
+    SecretNotFound,
+    SecretVersionNotFound,
+    VaultClient,
+    _VaultClient,
+)
+
+QONTRACT_INTEGRATION = "vault-replication"
+
+
+class VaultInvalidPaths(Exception):
+    pass
+
+
+class VaultInvalidAuthMethod(Exception):
+    pass
+
+
+def deep_copy_versions(
+    dry_run: bool,
+    source_vault: _VaultClient,
+    dest_vault: _VaultClient,
+    current_dest_version: int,
+    current_source_version: int,
+    path: str,
+) -> None:
+    for version in range(current_dest_version + 1, current_source_version + 1):
+        secret_dict = {"path": path, "version": version}
+
+        try:
+            secret, src_version = source_vault.read_all_with_version(secret_dict)
+            write_dict = {"path": path, "data": secret}
+            logging.info(["replicate_vault_secret", src_version, path])
+            if not dry_run:
+                dest_vault.write(
+                    secret=write_dict, decode_base64=False, force_write=True
+                )
+
+        except SecretNotFound:
```

Comment:
You have two calls in the `try` block, one to read, one to write. I would catch this only in one of the two calls so that it is clearer what you're trying.